distributed/utils.py

```diff
@@ -253,6 +256,112 @@ def f():
         return result[0]
 
 
+class LoopRunner(object):
+    """
+    A helper to start and stop an IO loop.
+
+    Parameters
+    ----------
+    loop: IOLoop (optional)
+        If given, this loop will be re-used, otherwise an appropriate one
+        will be looked up or created.
+    asynchronous: boolean (optional, default False)
+        If false (the default), the loop is meant to run in a separate
+        thread and will be started if necessary.
+        If true, the loop is meant to run in the thread this
+        object is instantiated from, and will not be started automatically.
+    """
+    def __init__(self, loop=None, asynchronous=False):
+        if loop is None:
+            if asynchronous:
+                self._loop = IOLoop.current()
+            else:
+                # We're expecting the loop to run in another thread,
+                # avoid re-using this thread's assigned loop
+                self._loop = IOLoop()
+            self._should_close_loop = True
+        else:
+            self._loop = loop
+            self._should_close_loop = False
+        self._asynchronous = asynchronous
+        self._loop_thread = None
+        self._started = False
+
+    def start(self):
+        """
+        Start the IO loop if required.  The loop is run in a dedicated
+        thread.
+
+        If the loop is already running, this method does nothing.
+        """
+        if self._asynchronous or self._loop_thread is not None:
+            self._started = True
+            return
+
+        loop_evt = threading.Event()
+        done_evt = threading.Event()
+        in_thread = [None]
+        start_exc = [None]
+
+        def loop_cb():
+            in_thread[0] = threading.current_thread()
+            loop_evt.set()
+
+        def run_loop(loop=self._loop):
+            loop.add_callback(loop_cb)
+            try:
+                loop.start()
+            except Exception as e:
+                start_exc[0] = e
+            finally:
+                done_evt.set()
+
+        thread = threading.Thread(target=run_loop,
+                                  name="IO loop")
+        thread.daemon = True
+        thread.start()
+
+        loop_evt.wait(timeout=1000)
+        self._started = True
+
+        actual_thread = in_thread[0]
+        if actual_thread is not thread:
+            # Loop already running in other thread (user-launched)
+            done_evt.wait(5)
+            if not isinstance(start_exc[0], RuntimeError):
+                raise start_exc[0]
+        else:
+            assert start_exc[0] is None, start_exc
+            self._loop_thread = thread
+
+    def stop(self, timeout=10):
+        """
+        Stop and close the loop if it was created by us.
+        Otherwise, just mark this object "stopped".
+        """
+        # XXX it would be more logical to stop an user-created loop
+        # if we started it, but that would break the current distributed API
```

Comment:
In which situations would this break user code?  Some breakage may be acceptable.

Reply:
Well, the problem is it currently breaks some of our tests. For example this one:
```python
def test_Current(loop):
    with cluster() as (s, [a, b]):
        with Client(s['address'], loop=loop) as c:
            assert Client.current() is c
        with pytest.raises(ValueError):
            Client.current()
        with Client(s['address'], loop=loop) as c:
            assert Client.current() is c
```

As soon as the first `with` end, the loop is closed, but it's passed again to the second `Client` instead, so the test hangs.

I don't know if there use cases for continuing to use a loop after you closed the client that started it, but this is how it currently works.

Reply:
If this is inconvenient then it might be reasonable to drop it.  The next version will likely be a 0.x.0 bump, so we can break a few things.

Reply:
It's not technically inconvenient, it just seems a bit inconsistent to have such assymetry in the start/stop responsabilities. But it probably wouldn't save any complexity to make it different.

Reply:
If you think it's a good idea then go for it.